// @progress/kendo-angular-inputs/dist/npm/slider/slider-ticks.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var kendo_inputs_common_1 = require("@telerik/kendo-inputs-common");
/**
 * @hidden
 */
var SliderTick = (function () {
    function SliderTick(value) {
        this.value = value;
        this.classes = {
            'k-tick': true
        };
    }
    return SliderTick;
}());
/* tslint:disable:component-selector */
/**
 * @hidden
 */
var SliderTicksComponent = (function () {
    function SliderTicksComponent(rtl) {
        this.rtl = rtl;
        this.wrapperClasses = 'k-reset k-slider-items';
        this.ticks = [];
    }
    SliderTicksComponent.prototype.ngOnChanges = function (_) {
        this.createTicks();
    };
    SliderTicksComponent.prototype.createTicks = function () {
        var count = kendo_inputs_common_1.SliderUtil.calculateTicksCount(this.max, this.min, this.step);
        var tickValueProps = {
            min: this.min,
            max: this.max,
            smallStep: this.step
        };
        var result = [];
        for (var i = 0; i < count; i++) {
            result.push(new SliderTick(kendo_inputs_common_1.SliderUtil.calculateValueFromTick(i, tickValueProps)));
        }
        if (this.rtl || this.vertical) {
            result = result.reverse();
        }
        if (result.length > 0) {
            Object.assign(result[0].classes, this.endTickClasses(true));
            Object.assign(result[result.length - 1].classes, this.endTickClasses(false));
        }
        this.ticks = result;
    };
    SliderTicksComponent.prototype.endTickClasses = function (first) {
        return {
            'k-first': (first && !this.vertical) || (!first && this.vertical),
            'k-last': (!first && !this.vertical) || (first && this.vertical)
        };
    };
    return SliderTicksComponent;
}());
SliderTicksComponent.decorators = [
    { type: core_1.Component, args: [{
                selector: '[kendoSliderTicks]',
                template: "\n    <li #tickElement *ngFor=\"let tick of ticks;\"\n        [ngClass]=\"tick.classes\"\n        title=\"{{ tickTitle(tick.value) }}\"\n        role=\"presentation\"\n     >&nbsp;</li>\n  "
            },] },
];
/** @nocollapse */
SliderTicksComponent.ctorParameters = function () { return [
    { type: undefined, decorators: [{ type: core_1.Optional }, { type: core_1.Inject, args: [kendo_angular_l10n_1.RTL,] },] },
]; };
SliderTicksComponent.propDecorators = {
    'wrapperClasses': [{ type: core_1.HostBinding, args: ['class',] },],
    'tickTitle': [{ type: core_1.Input },],
    'vertical': [{ type: core_1.Input },],
    'step': [{ type: core_1.Input },],
    'min': [{ type: core_1.Input },],
    'max': [{ type: core_1.Input },],
    'tickElements': [{ type: core_1.ViewChildren, args: ['tickElement',] },],
};
exports.SliderTicksComponent = SliderTicksComponent;
