// @progress/kendo-angular-dateinputs/dist/npm/calendar/calendar.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var common_1 = require("@angular/common");
var kendo_angular_intl_1 = require("@progress/kendo-angular-intl");
var templates_module_1 = require("./templates.module");
var calendar_component_1 = require("./calendar.component");
var header_component_1 = require("./header.component");
var navigation_component_1 = require("./navigation.component");
var view_list_component_1 = require("./view-list.component");
var view_component_1 = require("./view.component");
var bus_view_service_1 = require("./services/bus-view.service");
var dom_service_1 = require("./services/dom.service");
var navigation_service_1 = require("./services/navigation.service");
var century_view_service_1 = require("./services/century-view.service");
var decade_view_service_1 = require("./services/decade-view.service");
var month_view_service_1 = require("./services/month-view.service");
var year_view_service_1 = require("./services/year-view.service");
var localized_messages_directive_1 = require("./localization/localized-messages.directive");
var calendar_custom_messages_component_1 = require("./localization/calendar-custom-messages.component");
var virtualization_module_1 = require("../virtualization/virtualization.module");
var for_directive_1 = require("./for.directive");
var COMPONENT_DIRECTIVES = [
    calendar_component_1.CalendarComponent,
    header_component_1.HeaderComponent,
    navigation_component_1.NavigationComponent,
    calendar_custom_messages_component_1.CalendarCustomMessagesComponent,
    localized_messages_directive_1.LocalizedMessagesDirective,
    view_list_component_1.ViewListComponent,
    view_component_1.ViewComponent
];
var CALENDAR_PROVIDERS = [
    bus_view_service_1.BusViewService,
    dom_service_1.DOMService,
    navigation_service_1.NavigationService,
    century_view_service_1.CenturyViewService,
    decade_view_service_1.DecadeViewService,
    month_view_service_1.MonthViewService,
    year_view_service_1.YearViewService
];
/**
 * The exported package module.
 *
 * The package exports:
 * - `CalendarComponent`&mdash;The Calendar component class.
 * - `CellTemplateDirective`&mdash;The month cell template directive.
 * - `MonthCellTemplateDirective`&mdash;The month cell template directive.
 * - `YearCellTemplateDirective`&mdash;The year cell template directive.
 * - `DecadeCellTemplateDirective`&mdash;The decade cell template directive.
 * - `CenturyCellTemplateDirective`&mdash;The century cell template directive.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Calendar module
 * import { CalendarModule } from '@progress/kendo-angular-dateinputs';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * _@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, CalendarModule], // import Calendar module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var CalendarModule = (function () {
    function CalendarModule() {
    }
    return CalendarModule;
}());
CalendarModule.decorators = [
    { type: core_1.NgModule, args: [{
                declarations: [COMPONENT_DIRECTIVES, for_directive_1.KForOf],
                exports: [COMPONENT_DIRECTIVES, templates_module_1.TemplatesModule],
                imports: [common_1.CommonModule, kendo_angular_intl_1.IntlModule, templates_module_1.TemplatesModule, virtualization_module_1.VirtualizationModule],
                providers: CALENDAR_PROVIDERS
            },] },
];
/** @nocollapse */
CalendarModule.ctorParameters = function () { return []; };
exports.CalendarModule = CalendarModule;
