// @progress/kendo-angular-dateinputs/dist/npm/calendar/view.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* tslint:disable:component-selector-name  component-selector */
var core_1 = require("@angular/core");
var view_enum_1 = require("./models/view.enum");
var bus_view_service_1 = require("./services/bus-view.service");
var kendo_angular_intl_1 = require("@progress/kendo-angular-intl");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var util_1 = require("../util");
/**
 * @hidden
 */
var ViewComponent = (function () {
    function ViewComponent(bus, intl) {
        this.bus = bus;
        this.intl = intl;
        this.isActive = true;
        this.change = new core_1.EventEmitter();
        this.colSpan = 0;
        this.today = util_1.getToday();
    }
    Object.defineProperty(ViewComponent.prototype, "weekNumber", {
        get: function () {
            return this.showWeekNumbers && this.activeView === view_enum_1.CalendarViewEnum.month;
        },
        set: function (showWeekNumbers) {
            this.showWeekNumbers = showWeekNumbers;
        },
        enumerable: true,
        configurable: true
    });
    ViewComponent.prototype.ngOnChanges = function (_) {
        this.service = this.bus.service(this.activeView);
        if (!this.service) {
            return;
        }
        this.colSpan = this.service.rowLength(this.weekNumber);
        this.title = this.service.title(this.viewDate);
        var time = this.selectedDate || this.today;
        var viewDate = util_1.setTime(this.viewDate, time);
        this.data = this.service.data({
            cellUID: this.cellUID,
            focusedDate: this.focusedDate,
            isActiveView: !this.bus.canMoveDown(this.activeView),
            max: this.max,
            min: this.min,
            selectedDate: this.selectedDate,
            viewDate: viewDate
        });
    };
    ViewComponent.prototype.handleClick = function (ctx) {
        this.change.emit(ctx.value);
    };
    ViewComponent.prototype.firstDate = function (rowCtx) {
        var ctx = this.firstWeekDateContext(rowCtx);
        return ctx ? ctx.value : null;
    };
    ViewComponent.prototype.getWeekNumber = function (date) {
        if (!this.weekNumber) {
            return null;
        }
        return kendo_date_math_1.weekInYear(date, this.intl.firstDay());
    };
    ViewComponent.prototype.getWeekNumberContext = function (rowCtx) {
        var ctx = this.firstWeekDateContext(rowCtx);
        if (!this.weekNumber || !ctx) {
            return null;
        }
        var weekNumber = kendo_date_math_1.weekInYear(ctx.value, this.intl.firstDay()).toString();
        return {
            formattedValue: weekNumber,
            id: null,
            isFocused: false,
            isSelected: false,
            isWeekend: false,
            title: weekNumber,
            value: kendo_date_math_1.cloneDate(ctx.value)
        };
    };
    ViewComponent.prototype.getStyles = function (context) {
        return util_1.stringifyClassObject({
            'k-state-focused': this.isActive && context.isFocused,
            'k-state-selected': context.isSelected,
            'k-today': kendo_date_math_1.isEqual(context.value, this.today),
            'k-weekend': context.isWeekend
        });
    };
    ViewComponent.prototype.firstWeekDateContext = function (rowCtx) {
        if (!this.weekNumber) {
            return null;
        }
        var idx = 0;
        var ctx = rowCtx[idx];
        while (!ctx && idx < rowCtx.length) {
            ctx = rowCtx[++idx];
        }
        return ctx;
    };
    return ViewComponent;
}());
ViewComponent.decorators = [
    { type: core_1.Component, args: [{
                changeDetection: core_1.ChangeDetectionStrategy.OnPush,
                selector: '[kendoCalendarView]',
                template: "\n    <ng-template #emptyCell><td>&nbsp;</td></ng-template>\n    <tr role=\"row\"><th scope=\"col\" [colSpan]=\"colSpan\">{{title}}</th></tr>\n    <tr *kFor=\"let row of data\" role=\"row\">\n        <ng-template [ngIf]=\"weekNumber\">\n            <td class=\"k-alt\" *ngIf=\"firstDate(row); else emptyCell\">\n                <ng-template [ngIf]=\"!weekNumberTemplateRef\">\n                    {{getWeekNumber(firstDate(row))}}\n                </ng-template>\n                <ng-template\n                    [ngIf]=\"weekNumberTemplateRef\"\n                    [ngTemplateOutlet]=\"weekNumberTemplateRef\"\n                    [ngTemplateOutletContext]=\"{\n                        $implicit: firstDate(row),\n                        cellContext: getWeekNumberContext(row)\n                    }\"\n                ></ng-template>\n            </td>\n        </ng-template>\n        <ng-template kFor [kForOf]=\"row\" let-cell>\n            <td *ngIf=\"cell; else emptyCell\"\n                role=\"gridcell\" [attr.id]=\"cell.id\" [attr.aria-selected]=\"cell.isSelected\"\n                [ngClass]=\"getStyles(cell)\" [title]=\"cell.title\" (click)=\"handleClick(cell)\"\n            >\n                <span class=\"k-link\">\n                    <ng-template [ngIf]=\"!templateRef\">{{cell.formattedValue}}</ng-template>\n                    <ng-template\n                        [ngIf]=\"templateRef\"\n                        [ngTemplateOutlet]=\"templateRef\"\n                        [ngTemplateOutletContext]=\"{ $implicit: cell.value, cellContext: cell }\"\n                    ></ng-template>\n                </span>\n            </td>\n        </ng-template>\n    </tr>\n  "
            },] },
];
/** @nocollapse */
ViewComponent.ctorParameters = function () { return [
    { type: bus_view_service_1.BusViewService, },
    { type: kendo_angular_intl_1.IntlService, },
]; };
ViewComponent.propDecorators = {
    'isActive': [{ type: core_1.Input },],
    'activeView': [{ type: core_1.Input },],
    'cellUID': [{ type: core_1.Input },],
    'focusedDate': [{ type: core_1.Input },],
    'selectedDate': [{ type: core_1.Input },],
    'viewDate': [{ type: core_1.Input },],
    'min': [{ type: core_1.Input },],
    'max': [{ type: core_1.Input },],
    'weekNumber': [{ type: core_1.Input },],
    'viewIndex': [{ type: core_1.Input },],
    'templateRef': [{ type: core_1.Input },],
    'weekNumberTemplateRef': [{ type: core_1.Input },],
    'change': [{ type: core_1.Output },],
};
exports.ViewComponent = ViewComponent;
