// @progress/kendo-angular-grid/dist/npm/filtering/operators/after-eq-filter-operator.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var filter_operator_base_1 = require("./filter-operator.base");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
/*
 * Represents the `GreaterOrEqualTo` (**Is after or equal to**) date filter operator.
 */
var AfterEqFilterOperatorComponent = (function (_super) {
    tslib_1.__extends(AfterEqFilterOperatorComponent, _super);
    function AfterEqFilterOperatorComponent(localization) {
        return _super.call(this, "after-eq", localization) || this;
    }
    /**
     * @hidden
     */
    AfterEqFilterOperatorComponent.prototype.toJSON = function () {
        return {
            text: this.text,
            value: "gte"
        };
    };
    return AfterEqFilterOperatorComponent;
}(filter_operator_base_1.FilterOperatorBase));
AfterEqFilterOperatorComponent.decorators = [
    { type: core_1.Component, args: [{
                providers: [
                    {
                        provide: filter_operator_base_1.FilterOperatorBase,
                        useExisting: core_1.forwardRef(function () { return AfterEqFilterOperatorComponent; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-filter-after-eq-operator',
                template: ""
            },] },
];
/** @nocollapse */
AfterEqFilterOperatorComponent.ctorParameters = function () { return [
    { type: kendo_angular_l10n_1.LocalizationService, },
]; };
exports.AfterEqFilterOperatorComponent = AfterEqFilterOperatorComponent;
