// @progress/kendo-angular-grid/dist/npm/filtering/operators/is-not-empty-filter-operator.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var filter_operator_base_1 = require("./filter-operator.base");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
/**
 * Represents the `IsNotEmpty` (**Is not empty**) filter operator.
 */
var IsNotEmptyFilterOperatorComponent = (function (_super) {
    tslib_1.__extends(IsNotEmptyFilterOperatorComponent, _super);
    function IsNotEmptyFilterOperatorComponent(localization) {
        return _super.call(this, "isnotempty", localization) || this;
    }
    return IsNotEmptyFilterOperatorComponent;
}(filter_operator_base_1.FilterOperatorBase));
IsNotEmptyFilterOperatorComponent.decorators = [
    { type: core_1.Component, args: [{
                providers: [
                    {
                        provide: filter_operator_base_1.FilterOperatorBase,
                        useExisting: core_1.forwardRef(function () { return IsNotEmptyFilterOperatorComponent; }) // tslint:disable-line:no-forward-ref
                    }
                ],
                selector: 'kendo-filter-isnotempty-operator',
                template: ""
            },] },
];
/** @nocollapse */
IsNotEmptyFilterOperatorComponent.ctorParameters = function () { return [
    { type: kendo_angular_l10n_1.LocalizationService, },
]; };
exports.IsNotEmptyFilterOperatorComponent = IsNotEmptyFilterOperatorComponent;
