// @progress/kendo-angular-grid/dist/npm/filtering/menu/single-popup.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var kendo_angular_popup_1 = require("@progress/kendo-angular-popup");
var core_1 = require("@angular/core");
var utils_1 = require("../../utils");
var preventable_event_1 = require("../../common/preventable-event");
var Subject_1 = require("rxjs/Subject");
var kendo_angular_l10n_1 = require("@progress/kendo-angular-l10n");
var contains = function (node, predicate) {
    while (node) {
        if (predicate(node)) {
            return true;
        }
        node = node.parentNode;
    }
    return false;
};
var CloseEvent = (function (_super) {
    tslib_1.__extends(CloseEvent, _super);
    function CloseEvent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return CloseEvent;
}(preventable_event_1.PreventableEvent));
/**
 * @hidden
 */
var SinglePopupService = (function () {
    function SinglePopupService(popupService, renderer, ngZone, rtl) {
        if (rtl === void 0) { rtl = false; }
        this.popupService = popupService;
        this.renderer = renderer;
        this.ngZone = ngZone;
        this.rtl = rtl;
        this.onClose = new Subject_1.Subject();
    }
    SinglePopupService.prototype.open = function (anchor, template, popupRef) {
        var toggle = utils_1.isPresent(popupRef) && this.popupRef === popupRef;
        this.destroy();
        if (!toggle) {
            var direction = this.rtl ? 'right' : 'left';
            this.popupRef = this.popupService.open({
                anchorAlign: { vertical: 'bottom', horizontal: direction },
                popupAlign: { vertical: 'top', horizontal: direction },
                anchor: anchor,
                content: template
            });
            this.attachClose(anchor);
        }
        return this.popupRef;
    };
    SinglePopupService.prototype.destroy = function () {
        if (this.popupRef) {
            this.detachClose();
            this.popupRef.close();
            this.popupRef = null;
        }
    };
    SinglePopupService.prototype.detachClose = function () {
        if (this.removeClick) {
            this.removeClick();
        }
    };
    SinglePopupService.prototype.attachClose = function (skipElement) {
        var _this = this;
        this.detachClose();
        this.ngZone.runOutsideAngular(function () {
            return _this.removeClick = _this.renderer.listen("document", "click", function (_a) {
                var target = _a.target;
                if (!contains(target, function (x) { return _this.popupRef.popupElement === x || x === skipElement; })) {
                    var args = new CloseEvent();
                    _this.onClose.next(args);
                    if (!args.isDefaultPrevented()) {
                        _this.destroy();
                    }
                }
            });
        });
    };
    return SinglePopupService;
}());
SinglePopupService.decorators = [
    { type: core_1.Injectable },
];
/** @nocollapse */
SinglePopupService.ctorParameters = function () { return [
    { type: kendo_angular_popup_1.PopupService, },
    { type: core_1.Renderer2, },
    { type: core_1.NgZone, },
    { type: undefined, decorators: [{ type: core_1.Optional }, { type: core_1.Inject, args: [kendo_angular_l10n_1.RTL,] },] },
]; };
exports.SinglePopupService = SinglePopupService;
