// @progress/kendo-angular-grid/dist/npm/common/draggable.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_draggable_1 = require("@telerik/kendo-draggable");
/**
 * @hidden
 */
var DraggableDirective = (function () {
    function DraggableDirective(element, ngZone) {
        var _this = this;
        this.kendo = {
            drag: new core_1.EventEmitter(),
            press: new core_1.EventEmitter(),
            release: new core_1.EventEmitter()
        };
        if (typeof document !== 'undefined') {
            this.draggable = new kendo_draggable_1.default({
                drag: function (e) { return _this.kendo.drag.next(Object.assign({}, e, { column: _this.column })); },
                press: function (e) { return _this.kendo.press.next(Object.assign({}, e, { column: _this.column })); },
                release: function (e) { return _this.kendo.release.next(Object.assign({}, e, { column: _this.column })); }
            });
            ngZone.runOutsideAngular(function () { return _this.draggable.bindTo(element.nativeElement); });
        }
    }
    Object.defineProperty(DraggableDirective.prototype, "kendoGridDraggable", {
        set: function (column) {
            this.column = column;
        },
        enumerable: true,
        configurable: true
    });
    DraggableDirective.prototype.ngOnDestroy = function () {
        if (typeof document !== 'undefined') {
            this.draggable.destroy();
        }
    };
    return DraggableDirective;
}());
DraggableDirective.decorators = [
    { type: core_1.Directive, args: [{
                outputs: ['kendo.press', 'kendo.drag', 'kendo.release'],
                selector: '[kendoGridDraggable]'
            },] },
];
/** @nocollapse */
DraggableDirective.ctorParameters = function () { return [
    { type: core_1.ElementRef, },
    { type: core_1.NgZone, },
]; };
DraggableDirective.propDecorators = {
    'kendoGridDraggable': [{ type: core_1.Input },],
};
exports.DraggableDirective = DraggableDirective;
