// @progress/kendo-angular-dateinputs/dist/npm/timepicker/services/dom.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var util_1 = require("../../util");
var div = util_1.domContainerFactory('div');
var ul = util_1.domContainerFactory('ul');
var li = util_1.domContainerFactory('li');
var listItem = function () { return li('<span>02</span>', 'k-item'); };
var list = function () { return ul([listItem()], 'k-reset'); };
var scrollable = function () { return (div([list()], 'k-time-container k-content k-scrollable')); };
var timeListWrapper = function () {
    if (!util_1.isDocumentAvailable()) {
        return null;
    }
    return div([div([scrollable()], 'k-time-list')], 'k-time-list-wrapper', { left: '-10000px', position: 'absolute' });
};
var TIMELIST_WRAPPER = timeListWrapper();
/**
 * @hidden
 */
var DOMService = (function () {
    function DOMService() {
    }
    DOMService.prototype.ensureHeights = function () {
        if (this.timeListHeight !== undefined) {
            return;
        }
        this.calculateHeights();
    };
    DOMService.prototype.calculateHeights = function (container) {
        if (!util_1.isDocumentAvailable()) {
            return;
        }
        var listContainer = container && container.querySelector('.k-time-list-container');
        var hostContainer = listContainer || document.body;
        var wrapper = hostContainer.appendChild(TIMELIST_WRAPPER);
        this.timeListHeight = wrapper.querySelector('.k-scrollable').offsetHeight;
        this.itemHeight = wrapper.querySelector('li').offsetHeight;
        hostContainer.removeChild(wrapper);
    };
    DOMService.prototype.isActive = function (element) {
        if (!util_1.isDocumentAvailable() || !element) {
            return false;
        }
        return (element.nativeElement || element) === document.activeElement;
    };
    return DOMService;
}());
DOMService.decorators = [
    { type: core_1.Injectable },
];
/** @nocollapse */
DOMService.ctorParameters = function () { return []; };
exports.DOMService = DOMService;
